Fix relative import of base settings

roadsideseb opened the pull request:
I just tried out your template project for one of my own and got an import error for the `local.py`:

```
ImportError: Could not import settings 'synpho.settings.local' (Is it on sys.path?): No module named base
```

Looking at the files importing `base` it seems like the imports are supposed to be relative but are defined as absolute. Your code snippet in the book confirms that that's the intended way. I fixed it in the three settings files. I am a bit confused though since it seems that others have used the project successfully. So if I am missing something really obvious, just ignore this :smile:

Oh, and thanks for a great book. It's a great and educating read :thumbsup:

Commit:
Fix relative import of base settings

pratyushmittal commented:
The relative import works fine for me. I do `./manage.py runserver` from the project directory (where manage.py contains the hardcoded path to the `/project/settings/local`).

In the book too, on page 4, it is recommended to use relative paths within individual packages and apps.

pydanny commented:
I can't duplicate your problem. Can you detail where and how you called manage.py?

roadsideseb commented:
That's interesting, to reproduce the problem in a fresh project from your template, I can't reproduce it either. I'll take another look at my specific project and see what was happening there.

Regardless, I am not sure what I am missing because it seems to me that the imports in the settings files such as `from base import *` are absolute imports, right? How are they treated as relative then? Can you give me a hint where I am going blind? Thanks.

pydanny commented:
@elbaschid Where are you calling manage.py from? Can you give the shell commands? Otherwise I can't begin to figure out where we are doing things differently from each other. ;-)

pydanny commented:
Any response or should we close this as an unreproducable problem?

roadsideseb commented:
Sorry, I totally missed to reply to your response. I just tried a fresh setup following the readme and it all works fine, so I am not sure why it didn't work earlier. I'll close this until I can actually reproduce it. Until then, I'll consider it a bug of my own :wink: 

One last thing however, and I am sure I am missing something quite obvious: using `from base import *` is an absolute import but what happens in `local.py` is a relative import in the current directory which I would expect to be `from .base import *`. Can you enlighten me? What am I missing here?

pydanny commented:
Actually, I'm not certain. I just listen to the advice of giants. :wink: 

What I'm going to do is ask the giants. They are the ones who advocated absolute relative imports so strongly.

roadsideseb commented:
Ahhh, I was just able to reproduce this issue in a virtualenv with python 3.3. I know there might be other issues such as running `django-debug-toolbar` in a python 3 environment which doesn't have support for it yet. Would you mind checking if you can reproduce this with python 3 and let me know what you think about my changes suggested earlier?

roadsideseb commented:
We finally got an answer for this :smile: I had a chance to speak to the giants myself at PyConAU. And `from base import *` will **not** work in python 3.x, the correct way is `from .base import *`. Relative imports have to be explicit in python 3.x, all others are considered absolute paths. This is also explained in the [Python porting docs](http://docs.python.org/3/howto/pyporting.html#from-future-import-absolute-import). Thanks to the giants for clarifying this :smile:

pydanny commented:
Nice research! :smile: 

Here's the issue then: There are many users of this that are on Python 2.7 and don't have a problem. Changing to this pattern might cause problems for them and unfortunately I don't have the free time to do the correct amount of research. Therefore, I propose the following:
- Until the next edition of the book is released (January-ish, for Django 1.7), we keep this project as is.
- You maintain a Python 3 compatible fork. We'll link to it from the README.rst of this project.

Thoughts?

roadsideseb commented:
I completely understand your position on this and think this is a good way forward without potentially causing problems for people using 2.7. I am happy going doing this route. I'll try and merge my changes into my fork's master and add some documentation in the coming week. I'll ping you when it's done.

On a side note, I've come across an issue with implicit relative imports last week. In a Python 2.7 Django project, I had an app `awesome` and some other app that relied on an external module in 'site-packages' called `awesome`. This breaks things, because the absolute import of the `awesome` module is interpreted as implicit relative import of the `awesome` app. The only way around it is using `from __future__ import absolute_import` wherever the module is imported. I hope this makes sense. Does it make sense to add this to the documentation as a potential pitfall? Or is this too special a case?

pydanny commented:
Hey @elbaschid, want to maintain a Python 3 compat fork? I'll link to it from the README of this project.. If that's the case, I'll close this issue and point the README to your fork.

Commit:
Enforce explicit absolute imports

roadsideseb commented:
I just updated my fork and merged the branch in question into `master`. I also added a short disclaimer at the top of the README file referencing this conversation. Please take a look and let me know if that's alright or if you want me to change anything.

pmav99 commented:
I think it shouldn't be too difficult to solve this issue in a cross-version compatible way. Haven't tested it yet, but adding a from **future** import should do it.

``` python
from __future__ import absolute_imports

from .base import *
```

pydanny commented:
@pmav99 That's a pretty good idea! I'll see if I can try it out today.

In the meantime, I'm working on finally fixing the settings.py for the next evolution of this template, which is at: github.com/pydanny/cookiecutter-django

pmav99 commented:
Something like this should do it.

https://github.com/pmav99/django-twoscoops-project/commit/7908c3be88a9db265f367184d2b374c921f4170a

pydanny commented:
Hey @elbaschid, check out #65. If you want to update this, I'll accept it as a pull request.

roadsideseb commented:
That's great news :smile: I've rebased the PR and added the enforcement of only explicit absolute imports as mentioned by @pmav99. Does that look good?

pydanny commented:
Whoops - @elbaschid can you submit that rebase?

Argh. I think I might have screwed things up. Let me know!

roadsideseb commented:
I don't think you did. The merge seems fine but it seems that I've submitted my PR against `master` instead of `develop`. That's why the changes don't show up in `develop`. I hope that didn't cause you any issues. Let me know if there's anything I can do to help fix this.

pydanny commented:
Pushed it to `develop`. All is good. :dancer:

roadsideseb commented:
Awesome, thanks!
